examples/log.rs

```diff
@@ -0,0 +1,135 @@
+use crossterm::{
+    event::{self, DisableMouseCapture, EnableMouseCapture, Event, KeyCode},
+    execute,
+    terminal::{disable_raw_mode, enable_raw_mode, EnterAlternateScreen, LeaveAlternateScreen},
+};
+use log::{set_logger, set_max_level, warn, Log};
+use once_cell::sync::Lazy;
+use rand::random;
+use ratatui::{
+    backend::{Backend, CrosstermBackend},
+    text::Text,
+    widgets::{Block, Paragraph},
+    Frame, Terminal,
+};
+use std::{
+    error::Error,
+    fmt::Write,
+    io,
+    sync::Mutex,
+    time::{Duration, Instant},
+};
+
+struct App {}
+
+struct ParagraphLogger {
+    text: Mutex<Text<'static>>,
+}
+
+impl ParagraphLogger {
+    fn new() -> ParagraphLogger {
+        ParagraphLogger {
+            text: Mutex::new(Text::raw("")),
+        }
+    }
+}
+
+impl Log for ParagraphLogger {
+    fn enabled(&self, _: &log::Metadata) -> bool {
+        true
+    }
+
+    fn log(&self, record: &log::Record) {
+        writeln!(
+            self.text.lock().unwrap(),
+            "{}: {}",
+            record.level(),
+            record.args()
+        )
+        .unwrap();
+    }
+
+    fn flush(&self) {}
+}
+
+static LOGGER: Lazy<ParagraphLogger> = Lazy::new(ParagraphLogger::new);
+
+impl App {
+    fn new() -> App {
+        App {}
+    }
+
+    fn on_tick(&mut self) {
+        let rand: f32 = random();
+        if rand > 0.75 {
+            warn!("Tolerance exceeded!");
+        }
```

Comment:
Running this example, just splats out this message repeatedly. That makes it difficult to ascertain what's useful about this. Why is it more useful than just appending info to a string, or storing each log message in a `Vec`? What's the problem that this helps make easier to write? Going back to my benchmark comment, this example seems to save <1us every 250ms (i.e. it's a 0.0004% speedup over just using a String)

Reply:
My main objective is ergonomics, not performance—though if performance can be good too, that's a plus. Being able to `write!` directly to the `Text` is more convenient than collecting into an external buffer first. This would be especially true if the `Write` impl is able to parse escape sequences, in which case it will be possible to use crates like `crossterm::Stylize`:

```rust
// Example application code snippet
use crossterm::Stylize;
use log::Error;

fn report_error_with_foo(foo: &Foo) {
     error!("Oh no, an error occured! {}", foo).bold().red();
}
```

It would also be useful for building up a `Text` more directly. For example:

```rust
let mut text = Text::raw("");
writeln!(text, "You have {} new messages!", num_messages.bold().blue());
write!(text, "Last refresh: {}", refresh_time.underlined().gray());
// Display `text` inside paragraph widget...
```

Reply:
Yup - I definitely agree that it is good idea for those reasons. I want to see your good idea reflected in the code though so that the rationale isn't just floating around in a PR that no-one looks at again once it's merged. And while I think it's a good idea, it has some problems.

I'm a little hesitant about what to do with the ANSI codes though. I'd want to see how the ergonomics of that turn out.

Re:

> Being able to write! directly to the Text is more convenient than collecting into an external buffer first.

How much different is the buffer vs text though? I'd have thought it's two lines of code.

I think my main concern on this PR is that it constrains us to very specific handling of lines and continuations of the lines, where previously we had a notion of these being a string per line created whenever you render, now we have to worry about that state being maintained, with an indicator of whether a line is terminated by a newline or not. It's a really subtle change in behavior done implicitly rather than explicitly.

One example of the weirdness of this (ergonomically) is that `write!` becomes the only mechanism to extend a line - there doesn't seem like there's a corresponding `extend_line` method. Text only deals with lines, except when you're writing, then it deals in partial lines, but the line state is not something that a caller can inspect / understand without reading the implementation.